### PR TITLE
20 mutual information

### PR DESCRIPTION
**Pull request by QianyeYang**
# Description

Please include a summary of the change and which issue is fixed. Please also include
relevant motivation and context. List any dependencies that are required for this
change.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not
      work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so
we can reproduce. Please also list any relevant details for your test configuration If
there are no tests written, flag with label so we can help generate some tests before
merge.

- [ ] Test A
- [ ] Test B

**Test Configuration**:

- Firmware version:
- Hardware:

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: test/unit/test_loss_image.py [hunk 1]
```diff
@@ -55,6 +55,11 @@ def test_dissimilarity_fn():
     )
     assert is_equal_tf(get_zero_similarity_ssd, [0, 0])
 
+    #  testing if we can get [0, 0] (i.e. zero vector) in mutual information
+    t = tf.ones([4, 3, 3, 3, 1])
+    get_zero_similarity_gmi = image.dissimilarity_fn(t, t, "gmi")
+    assert tf.reduce_max(tf.abs(get_zero_similarity_gmi - [0, 0, 0, 0])).numpy() < 5e-6
```
Comment (mathpluscode) [inline test/unit/test_loss_image.py:61]: @QianyeYang why we can't use the `is_equal_tf` here?

File: deepreg/model/loss/image.py [hunk 2]
```diff
@@ -82,3 +84,43 @@ def ssd(y_true: tf.Tensor, y_pred: tf.Tensor) -> tf.Tensor:
...
+    I_a /= tf.reduce_sum(I_a, -1, keepdims=True)
+    pa = tf.reduce_mean(I_a, axis=1, keepdims=True)
+    I_a_permute = tf.transpose(I_a, (0, 2, 1))
+
+    I_b = tf.math.exp(-preterm * tf.math.square(y_pred - vbc))
+    I_b /= tf.reduce_sum(I_b, -1, keepdims=True)
+    pb = tf.reduce_mean(I_b, axis=1, keepdims=True)
+
+    pa = tf.transpose(pa, (0, 2, 1))
+    papb = tf.keras.backend.batch_dot(pa, pb, axes=(2, 1)) + eps
```
Comment (mathpluscode) [inline deepreg/model/loss/image.py:122]: maybe we can use pa * pb directly? the dot product should be boardcastable

File: deepreg/model/loss/image.py [hunk 3]
```diff
@@ -82,3 +84,43 @@ def ssd(y_true: tf.Tensor, y_pred: tf.Tensor) -> tf.Tensor:
     :return: shape = (batch,)
     """
     return tf.reduce_mean(tf.square(y_true - y_pred), axis=[1, 2, 3, 4])
+
+
+def global_mutual_information(y_true: tf.Tensor, y_pred: tf.Tensor) -> tf.Tensor:
```
Comment (mathpluscode) [inline deepreg/model/loss/image.py:89]: I personally would like more comments on the math and the shapes of the tensors to help people to understand this function.

File: deepreg/model/loss/image.py [hunk 4]
```diff
@@ -82,3 +84,43 @@ def ssd(y_true: tf.Tensor, y_pred: tf.Tensor) -> tf.Tensor:
...
+    """
+    differentiable global mutual information loss via Parzen windowing method.
+    reference: https://dspace.mit.edu/handle/1721.1/123142
```
Comment (mathpluscode) [inline deepreg/model/loss/image.py:92]: this reference is relatively long, is there any specific equation we can use as a reference?

File: deepreg/model/loss/image.py (hunk 2 above)
Comment (YipengHu) [reply deepreg/model/loss/image.py:122]: agree. avoid using backend functions if possible which caused compatibility issues before...

File: deepreg/model/loss/image.py (hunk 3 above)
Comment (YipengHu) [reply deepreg/model/loss/image.py:89]: > I personally would like more comments on the math and the shapes of the tensors to help people to understand this function.

ideally yes - but this should not be the reason for refuting the merge, as more detailed technical reference has already been provided.

File: test/unit/test_loss_image.py (hunk 1 above)
Comment (YipengHu) [reply test/unit/test_loss_image.py:61]: I actually would add that the use of consistent asserting functions is a good practice for future maintenance of the tests - for example, when the machine precision changes...

Comment (QianyeYang) [reply test/unit/test_loss_image.py:61]: @mathpluscode Because the EPS in that function is defined as 1e-6, but the value here is 2e-6.

File: deepreg/model/loss/image.py (hunk 4 above)
Comment (mathpluscode) [reply deepreg/model/loss/image.py:92]: also i found the code is a bit similar compared to https://github.com/voxelmorph/voxelmorph/blob/legacy/src/losses.py#L352

if we are referring to some code, maybe we should put a link in docstring?
(like if there's a bug later, it's because of the source lol)

Comment (mathpluscode) [reply deepreg/model/loss/image.py:92]: btw, just to be fair, the lncc loss I wrote earlier was not a good example neither lol
will improve it

File: deepreg/model/loss/image.py (hunk 2 above)
Comment (QianyeYang) [reply deepreg/model/loss/image.py:122]: Personally I prefer not to change if it's not necessary. Use a function here will make the purpose clear, and the readers don't need to think about the shape after broadcasting.

File: deepreg/model/loss/image.py (hunk 3 above)
Comment (QianyeYang) [reply deepreg/model/loss/image.py:89]: Will add the shape of the tensors and show the detailed pages and equations of the pdf. The corresponding part is easy to follow, so I'll just make some short comments on math if it's necessary.

File: deepreg/model/loss/image.py (hunk 4 above)
Comment (QianyeYang) [reply deepreg/model/loss/image.py:92]: As the above one.

File: test/unit/test_loss_image.py (hunk 1 above)
Comment (mathpluscode) [reply test/unit/test_loss_image.py:61]: @QianyeYang This is a bit strange? as we are expecting zero values....
maybe for some division you need to add eps for both denominator and numerator

File: deepreg/model/loss/image.py (hunk 2 above)
Comment (QianyeYang) [reply deepreg/model/loss/image.py:122]: > 
> 
> agree. avoid using backend functions if possible which caused compatibility issues before...

all right, it have to be changed if compatible issues exist.

Comment (mathpluscode) [reply deepreg/model/loss/image.py:122]: acc to https://stackoverflow.com/questions/59361689/redundancies-in-tf-keras-backend-and-tensorflow-libraries
although not many votes, i think it's better to not use backend whenever possible ;)
if you want, leave it there and i can change them tmr

File: deepreg/model/loss/image.py (hunk 4 above)
Comment (QianyeYang) [reply deepreg/model/loss/image.py:92]: > 
> 
> also i found the code is a bit similar compared to https://github.com/voxelmorph/voxelmorph/blob/legacy/src/losses.py#L352
> 
> if we are referring to some code, maybe we should put a link in docstring?
> (like if there's a bug later, it's because of the source lol)

sure, the method is from the same paper and I use their code as a reference. Will add the link if necessary.

Comment (mathpluscode) [reply deepreg/model/loss/image.py:92]: @QianyeYang not necessary if you didn't use them directly lol Im saying this just bcz sometimes I will take a look as existing code and make a comparison to validate my code ;)

File: deepreg/model/loss/image.py (hunk 2 above)
Comment (QianyeYang) [reply deepreg/model/loss/image.py:122]: > 
> 
> acc to https://stackoverflow.com/questions/59361689/redundancies-in-tf-keras-backend-and-tensorflow-libraries
> although not many votes, i think it's better to not use backend whenever possible ;)
> if you want, leave it there and i can change them tmr

I can fix them, will tell you if I need your help, many thanks!

File: test/unit/test_loss_image.py (hunk 1 above)
Comment (QianyeYang) [reply test/unit/test_loss_image.py:61]: > 
> 
> @QianyeYang This is a bit strange? as we are expecting zero values....
> maybe for some division you need to add eps for both denominator and numerator

will have a try adding eps on numerator, thanks!
